extensions/iceberg/src/main/java/io/deephaven/iceberg/util/IcebergTableWriter.java

```diff
@@ -75,191 +77,181 @@ public class IcebergTableWriter {
+     */
+    private final OutputFileFactory outputFileFactory;
 
-    private OutputFileFactory outputFileFactory;
 
     IcebergTableWriter(
             final TableWriterOptions tableWriterOptions,
             final IcebergTableAdapter tableAdapter) {
-        this.tableDefinition = tableWriterOptions.tableDefinition();
-        this.dhColumnsToIcebergFieldIds = tableWriterOptions.dhColumnsToIcebergFieldIds();
-        this.icebergFieldIdToDhColumn = new HashMap<>();
-        this.table = tableAdapter.icebergTable();
-        this.userSchema = tableWriterOptions.schema().orElseGet(table::schema);
-        verifyFieldIds(dhColumnsToIcebergFieldIds.values(), userSchema);
         this.tableAdapter = tableAdapter;
+        this.table = tableAdapter.icebergTable();
 
         if (table instanceof HasTableOperations) {
             tableMetadata = ((HasTableOperations) table).operations().current();
         } else {
             tableMetadata = null;
         }
-        inferSpecAndSchemaToWrite();
-    }
 
-    /**
-     * Check that all the field IDs are present in the schema.
-     */
-    private void verifyFieldIds(final Collection<Integer> fieldIds, final Schema schema) {
-        if (!fieldIds.isEmpty()) {
-            for (final Integer fieldId : fieldIds) {
-                if (schema.findField(fieldId) == null) {
-                    throw new IllegalArgumentException("Column corresponding to field ID " + fieldId + " not " +
-                            "found in schema, available columns in schema are: " + schema.columns());
-                }
+        this.tableDefinition = tableWriterOptions.tableDefinition();
+
+        // We only write non-partitioning columns to parquet file
+        this.parquetTableDefinition = TableDefinition.of(tableDefinition.getColumnStream()
+                .filter(columnDefinition -> !columnDefinition.isPartitioning())
+                .collect(Collectors.toList()));
+
+        this.userSchema = tableWriterOptions.schemaSpec().getSchema(table);
+
+        outputFileFactory = OutputFileFactory.builderFor(table, 0, 0)
+                .format(FileFormat.PARQUET)
+                .build();
+
+        // Create a copy of the fieldIdToColumnName map since we might need to add new entries for columns which are not
+        // present in the schema.
+        this.fieldIdToColumnName = new HashMap<>(tableWriterOptions.fieldIdToColumnName());
+        verifyFieldIds(fieldIdToColumnName.keySet(), userSchema);
+
+        // Populate the fieldIdToColumnName map for all the columns in the table definition and do additional checks
+        // to ensure that the table definition is compatible with the existing table.
+        {
+            final int lastColumnId;
+            if (tableMetadata != null) {
+                lastColumnId = tableMetadata.lastColumnId();
+            } else {
+                lastColumnId = table.schema().highestFieldId();
```

Comment:
So, I get what we are doing here; but these "inferred" field IDs are useless, and not guaranteed to be the same sort of thing that the implementation would use.

By the time we get to this writing stage, we already need to know "we are good to go".

For example, in the case where we need to support updating a schema and writing the table within a single transaction, it will need to look roughly like:

```java
// Create an updateSchema according to TableDefinition vs user provided schema spec
UpdateSchema updateSchema = table.updateSchema()
        .addColumn("Foo", Types.LongType.get())
        .renameColumn("Bar", "Baz");

final Transaction txn = table.newTransaction();

// This is the only way to correctly get the field-ids in Iceberg transaction-safe way
Schema newSchema = updateSchema.apply();
updateSchema.commit();

final IcebergTableWriter writer = new IcebergTableWriter(
        TableWriterOptions.builder().schemaSpec(SchemaSpec.fromSchema(newSchema)).build(), tableAdapter);

writer.append(...);

// use new schema for writer
txn.commitTransaction();
```